Issue #2668 - Delete a specified context

masa213f opened the pull request:
Checklist:

* [x] (b) this is a bug fix
* [x] The title of the PR states what changed and the related issues number (used for the release note).
* [x] I've signed the CLA and my build is green ([troubleshooting builds](https://argoproj.github.io/argo-cd/developer-guide/ci/)). 
---

Fix for #2668. Please review.

This code behaves as follows:
```shell
# 3 contexts exists.
$ argocd context
CURRENT  NAME                     SERVER
         argocd.hoge.example.com  argocd.hoge.example.com
*        argocd.piyo.example.com  argocd.piyo.example.com
         argocd.fuga.example.com  argocd.fuga.example.com

# Delete a non-current context.
$ argocd context --delete argocd.hoge.example.com
Context 'argocd.hoge.example.com' deleted
$ argocdq context
CURRENT  NAME                     SERVER
*        argocd.piyo.example.com  argocd.piyo.example.com
         argocd.fuga.example.com  argocd.fuga.example.com

# If a context name is not specified, delete the current context.
$ argocd context --delete
Context 'argocd.piyo.example.com' deleted
$ argocd context
CURRENT  NAME                     SERVER
*        argocd.fuga.example.com  argocd.fuga.example.com

# Even if the current context is specified, it is of course deleted.
$ argocd context --delete argocd.fuga.example.com
Context 'argocd.fuga.example.com' deleted
$ argocd context
FATA[0000] No contexts defined in /home/masa213f/.argocd/config
```

alexec commented:
Any thoughts about how we write tests for this?

jannfis commented:
> Any thoughts about how we write tests for this?

I think CLI tests should all be in end-to-end tests. The context deletion core functionality already has tests in `cmd/argocd/commands/context_test.go`, but we are not assured a good build (binary) of `argocd` when running `make test`.

masa213f commented:
@jannfis 
Thank you for your review.

Certainly, the implicit switch by the deletion of the current context may cause manipulating the unexpected cluster.

In my opinion, to reduce this risk, `argocd` command should not select the next current-context after the deletion and should request the explicit context change. 

For example:
```shell
# 3 contexts exists.
$ argocd context
CURRENT  NAME                     SERVER
         argocd.hoge.example.com  argocd.hoge.example.com
*        argocd.piyo.example.com  argocd.piyo.example.com
         argocd.fuga.example.com  argocd.fuga.example.com

# If the current context is deleted...
$ argocd context --delete argocd.piyo.example.com
Context 'argocd.hoge.example.com' deleted

# The current context will not be set. And some subcommands will be error. 
$ argocd context
CURRENT  NAME                     SERVER
         argocd.hoge.example.com  argocd.hoge.example.com
         argocd.fuga.example.com  argocd.fuga.example.com
$ argocd app list
=> Cause an error and request a context switch by the error message.
```

masa213f commented:
And I thought the context should only be deleted if it is given explicitly.
Because `argocd logout` subcommand which has a similar role requires the target context explicitly.

masa213f commented:
If the above specifications are okay, I will try it. Please wait.

alexmt commented:
@masa213f , thank you for PR!

I think the behavior that you described in this [comment](https://github.com/argoproj/argo-cd/pull/2669#issuecomment-552100364) is correct. The context should only be deleted if it is given explicitly.

masa213f commented:
@jannfis @alexmt 
Thank you for the comment.

I updated the context deletion. Please review.
I’m going to revise the code to reflect the changes whatever as you pointed out. :)

Adding end-to-end tests is difficult for me. So I only updated unit tests.

```shell
# 3 contexts exists.
$ argocd context
CURRENT  NAME                     SERVER
         argocd.hoge.example.com  argocd.hoge.example.com
*        argocd.piyo.example.com  argocd.piyo.example.com
         argocd.fuga.example.com  argocd.fuga.example.com

# The deletion will fail and print the usage if a context name is not given.
$ argocd context --delete
Switch between contexts

Usage:
  argocd context [CONTEXT] [flags]

(...)

# If the current context is deleted, the implicit context switch will not happen.
$ argocd --delete argocd.piyo.example.com
Context 'argocd.piyo.example.com' deleted
$ argocd context
CURRENT  NAME                     SERVER
         argocd.hoge.example.com  argocd.hoge.example.com
         argocd.fuga.example.com  argocd.fuga.example.com
$ argocd app list
FATA[0000] Local config: current-context unset

# In this case, users need an explicit context switch.
$ argocd context argocd.fuga.example.com
Switched to context 'argocd.fuga.example.com'
$ argocd context
CURRENT  NAME                     SERVER
         argocd.hoge.example.com  argocd.hoge.example.com
*        argocd.fuga.example.com  argocd.fuga.example.com
```
